### PR TITLE
[opencl][quantization]: Support quantization for OCLPoolAvg

### PR DESCRIPTION
**Pull request by ZchiPitt**
Support quantization for OCLPoolAvg on OpenCL backend

### CODE REVIEW

**Comment by rdzhabarov** [comment]
+1 to opti-mix, current implementation in this PR is not correct. Look at other already in place implementations.

**Comment by ZchiPitt** [comment]
Done with the improvement.  The test can pass previously, was because I only checked the value in the handle, which does take offset into consideration.

File: lib/Backends/OpenCL/kernels.cl [hunk 1]
```diff
@@ -1141,6 +1141,56 @@ __kernel void oclavgpoolW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
   oclavgpoolK(&mem[dest], &mem[src], filterSize, stride, pads, odim, idim);
 }
 
+__kernel void oclavgpool_i8K(__global cl_int8_t *dest, __global cl_int8_t *src,
```
Comment (rdzhabarov) [inline lib/Backends/OpenCL/kernels.cl]: why there is this "ocl" prefix?

File: lib/Backends/OpenCL/kernels.cl [hunk 2]
```diff
@@ -1141,6 +1141,56 @@ __kernel void oclavgpoolW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
...
+      }
+    }
+    // All dest scale params are already divided by kernel*kernel(filter area),
+    // we don't need to divide the sumVal by filter area here.
+    dest[getNCHW(odim, n, d, ax, ay)] = 
+        clip(scale_i32i8(sumVal, destPre, destPost, destScale, destOffset));
+  }
+}
+
+__kernel void oclavgpool_i8W(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
```
Comment (rdzhabarov) [inline lib/Backends/OpenCL/kernels.cl]: ditto

File: tests/unittests/OperatorTest.cpp [hunk 3]
```diff
@@ -2367,6 +2367,22 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
   EXPECT_TRUE(result.isEqual(result1));
 }
 
+TEST_P(Operator, Int8Poolavg) {
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: Int8AvgPool (there was some renaming of maxpool and avgpool)

File: tests/unittests/OperatorTest.cpp [hunk 4]
```diff
@@ -2367,6 +2367,22 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
...
+  input->getHandle<int8_t>() = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+  auto *Pool = F_->createAvgPool("pool", input, 2, 1, {0, 0, 0, 0});
+  auto *S = F_->createSave("save", Pool);
+  EE_.compile(CompilationMode::Infer, F_);
+  EE_.run({}, {});
+  auto result = S->getVariable()->getHandle<int8_t>();
+  auto *output = mod_.createVariable(ElemKind::Int8QTy, {2, 2}, 1, 0, "output");
+  output->getHandle<int8_t>() = {2, 3, 5, 6};
+  for (size_t i = 0; i < 2 * 2; i++) {
+    EXPECT_TRUE(result.raw(i) == output->getHandle<int8_t>().raw(i));
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: nit: EXPECT_EQ( .., ..)

File: tests/unittests/OperatorTest.cpp [hunk 5]
```diff
@@ -2367,6 +2367,22 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
...
+  }
+}
+
 TEST_P(Operator, Int8Poolmax) {
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: can you rename this as well?

File: tests/unittests/OperatorTest.cpp [hunk 6]
```diff
@@ -2367,6 +2367,22 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
...
+TEST_P(Operator, Int8Poolavg) {
+  auto *input =
+      mod_.createVariable(ElemKind::Int8QTy, {1, 3, 3, 1}, 1, 0, "input");
+  input->getHandle<int8_t>() = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+  auto *Pool = F_->createAvgPool("pool", input, 2, 1, {0, 0, 0, 0});
+  auto *S = F_->createSave("save", Pool);
+  EE_.compile(CompilationMode::Infer, F_);
+  EE_.run({}, {});
+  auto result = S->getVariable()->getHandle<int8_t>();
+  auto *output = mod_.createVariable(ElemKind::Int8QTy, {2, 2}, 1, 0, "output");
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: why do you need to create a var for this?

File: lib/Backends/OpenCL/kernels.cl (hunk 1 above)
Comment (ZchiPitt) [reply lib/Backends/OpenCL/kernels.cl]: it's the int8 version of kernel oclavgpoolK, and the instruction being executed here is also OCLAvgPoolInst. Not sure why we have both OCLAvgPool and AvgPool? @opti-mix

File: tests/unittests/OperatorTest.cpp (hunk 3 above)
Comment (ZchiPitt) [reply tests/unittests/OperatorTest.cpp]: renamed

File: tests/unittests/OperatorTest.cpp (hunk 4 above)
Comment (ZchiPitt) [reply tests/unittests/OperatorTest.cpp]: done

File: tests/unittests/OperatorTest.cpp (hunk 6 above)
Comment (ZchiPitt) [reply tests/unittests/OperatorTest.cpp]: changed to use a tensor

File: lib/Backends/OpenCL/kernels.cl (hunk 1 above)
Comment (rdzhabarov) [reply lib/Backends/OpenCL/kernels.cl]: Just talked to @opti-mix about this. Seems like it has to do with the tensor shape (NHWC vs NCHW). Would be good to rename these methods accordingly, if you prefer on a separate cleanup review.

Comment (ZchiPitt) [reply lib/Backends/OpenCL/kernels.cl]: sounds cool. But as it would touch more than these files we have in this pr, maybe we'll do in another pr?

**Comment by ZchiPitt** [comment]
resolved with the conflicts now. Prob i can merge this pr and rename or format the things in the following pr

File: lib/Backends/OpenCL/kernels.cl (hunk 1 above)
Comment (rdzhabarov) [reply lib/Backends/OpenCL/kernels.cl]: sounds good.

File: lib/Backends/OpenCL/kernels.cl [hunk 7]
```diff
@@ -1141,6 +1141,56 @@ __kernel void oclavgpoolW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
...
+
+  typedef int ssize_t;
+  // For each convolution 'jump' in the input tensor:
+  ssize_t x = -(ssize_t)pads.top + ax * stride;
+  ssize_t y = -(ssize_t)pads.left + ay * stride;
+
+  // For each input in the batch:
+  for (size_t n = 0; n < idim.n; n++) {
+    cl_int32_t sumVal = 0;
+    // For each element in the convolution-filter:
```
Comment (rdzhabarov) [inline lib/Backends/OpenCL/kernels.cl]: that's not a convolution filter ;)

File: lib/Backends/OpenCL/kernels.cl [hunk 8]
```diff
@@ -1141,6 +1141,56 @@ __kernel void oclavgpoolW(__global void *mem, cl_uint32_t dest, cl_uint32_t src,
...
+                       PaddingTLBR pads, ShapeNCHW odim, ShapeNCHW idim,
+                       cl_int32_t srcOffset, cl_int32_t destOffset,
+                       cl_int32_t destPre, cl_int32_t destPost, 
+                       cl_int32_t destScale) {
+  size_t ax = get_global_id(0);
+  size_t ay = get_global_id(1);
+  size_t d = get_global_id(2);
+
+  typedef int ssize_t;
+  // For each convolution 'jump' in the input tensor:
```
Comment (rdzhabarov) [inline lib/Backends/OpenCL/kernels.cl]: ditto

File: tests/unittests/OperatorTest.cpp [hunk 9]
```diff
@@ -2367,10 +2367,25 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
   EXPECT_TRUE(result.isEqual(result1));
 }
 
-TEST_P(Operator, Int8Poolmax) {
+TEST_P(Operator, Int8Avgpool) {
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: AvgPool

File: tests/unittests/OperatorTest.cpp [hunk 10]
```diff
@@ -2367,10 +2367,25 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
...
+  EE_.run({}, {});
+  auto result = S->getVariable()->getHandle<int8_t>();
+  Tensor out(ElemKind::Int8QTy, {2, 2}, 1, 0);
+  out.getHandle<int8_t>() = {2, 3, 5, 6};
+  for (size_t i = 0; i < 2 * 2; i++) {
+    EXPECT_EQ(result.raw(i), out.getHandle<int8_t>().raw(i));
+  }
+}
+
+TEST_P(Operator, Int8Maxpool) {
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: MaxPool

File: tests/unittests/OperatorTest.cpp [hunk 11]
```diff
@@ -2367,10 +2367,25 @@ TEST_P(Operator, NonSquarePaddingMaxPool) {
...
   auto *input =
       mod_.createVariable(ElemKind::Int8QTy, {1, 3, 3, 1}, 1, 0, "input");
   input->getHandle<int8_t>() = {0, 1, 2, 3, 4, 5, 6, 7, 8};
+  auto *Pool = F_->createAvgPool("pool", input, 2, 1, {0, 0, 0, 0});
+  auto *S = F_->createSave("save", Pool);
+  EE_.compile(CompilationMode::Infer, F_);
+  EE_.run({}, {});
+  auto result = S->getVariable()->getHandle<int8_t>();
+  Tensor out(ElemKind::Int8QTy, {2, 2}, 1, 0);
+  out.getHandle<int8_t>() = {2, 3, 5, 6};
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: `handle` has equals method, do not need to iterate here explicitly.

File: lib/Backends/OpenCL/kernels.cl (hunk 7 above)
Comment (ZchiPitt) [reply lib/Backends/OpenCL/kernels.cl]: removed

**Pull request merged**